pkg/controller/combo_controller.go

```diff
@@ -2,36 +2,134 @@ package controller
+	if err != nil {
 		return reconcile.Result{}, client.IgnoreNotFound(err)
 	}
 
+	// If combination is being deleted, remove from queue
+	if !combination.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("combination is being deleted, ignoring event")
+		return reconcile.Result{}, nil
+	}
+
+	log.Info(fmt.Sprintf("combination %s successfully found in reconciler", combination.Name))
+
+	// Attempt to retrieve the template referenced in the combination CR
+	template := &v1alpha1.Template{}
+	templateQuery := types.NamespacedName{Name: combination.Spec.Template}
+	if err := c.Get(ctx, templateQuery, template); err != nil {
+		err = fmt.Errorf("failed to retrieve %v template: %w", combination.Spec.Template, err)
+		return c.manageFailure(ctx, combination, comboConditions.TemplateNotFoundCondition, err)
+	}
+
+	// Build combination stream to be utilized in template builder
+	comboStream := combinationPkg.NewStream(
+		combinationPkg.WithArgs(formatArguments(combination.Spec.Arguments)),
+		combinationPkg.WithSolveAhead(),
+	)
+
+	// Create a new template builder
+	builder, err := templatePkg.NewBuilder(strings.NewReader(template.Spec.Body), comboStream)
+	if err != nil {
+		err = fmt.Errorf("failed to construct a builder out of %s template body: %w", template.Name, err)
+		return c.manageFailure(ctx, combination, comboConditions.TemplateBodyInvalid, err)
+	}
+
+	// Build the manifest combinations
+	generatedManifests, err := builder.Build(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to generate manifest %s combinations: %w", combination.Name, err)
+		return c.manageFailure(ctx, combination, comboConditions.ManifestGenerationFailed, err)
+	}
+
+	log.Info(fmt.Sprintf("reconciliation of %s combination complete!", combination.Name))
+
+	// Return and update the combination's status
+	return c.manageSuccess(ctx, combination, generatedManifests)
+}
+
+// manageSuccess takes a combination CR and its evaluation to process a successful status update for it
+func (c *combinationController) manageSuccess(ctx context.Context, combination *v1alpha1.Combination, evaluation []string) (reconcile.Result, error) {
+	// Create the new status condition to transition to
+	combination.Status.Conditions = comboConditions.NewConditions(time.Now(), nil, SuccessCondition)
+
+	combination.Status.Evaluation = evaluation
+
+	// Update the status of the combination, requeue if this update fails
+	updateErr := c.Status().Update(ctx, combination)
+	if updateErr != nil {
+		c.log.Info("Error when updating success status, requeuing: " + updateErr.Error())
+		return reconcile.Result{RequeueAfter: RequeueDefaultTime}, updateErr
+	}
+
 	return reconcile.Result{}, nil
 }
+
+// manageFailure takes a combination CR, the new condition and whatever error occurred to process a failure status update for it
+func (c *combinationController) manageFailure(ctx context.Context, combination *v1alpha1.Combination, condition metav1.Condition, err error) (reconcile.Result, error) {
```

Comment:
Any reason we couldn't deflate this set condition logic + update calls at the actual call site? I'm sure I see a ton of value in this type of function, but I'm open to any thoughts/opinions/etc.

Reply:
I think that it makes the code a bit cleaner in the actual Reconciler function. It makes it so we're repeating less code in the err checks (such as line 84-87 of this file) plus, we don't need to keep rewriting the "updateErr" check for the update itself. Would the alternative of deflating it to the original call site be to just do this in each err check instead?

Reply:
This is essentially what I had in mind:

```go
if err != nil {
    combination.SetSomeCondition(...)
    // maybe wrap the err here as well with any error that the status sub-resource call produces?
    return ctrl.Result{}, c.Client.Status().Update(ctx, combination)
```

Reply:
Yeah, I personally don't mind it factored out. But, there's a pretty big overlap between `manageFailure` and `manageSuccess`. I would just merge it into one called `updateStatusAndReturn` or something like that. Then just pass nil for err in the success case.

Reply:
How do we feel about this change?

https://github.com/operator-framework/combo/commit/477e6230dcc3d4d27435329c4751d40c7bfb5f81